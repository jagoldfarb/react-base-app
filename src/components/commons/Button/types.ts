import { ReactNode } from 'react';

export interface IButtonProps {
  children: ReactNode;
  className?: string;
  disabled?: boolean;
  href?: string;
  onClick?: () => void;
  type?: 'button' | 'submit' | 'reset' | undefined;
}
