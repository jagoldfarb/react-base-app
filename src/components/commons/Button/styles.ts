import styled from 'styled-components';

import { COLORS } from 'helpers/enums/colors';

export const CustomButton = styled.button`
  border: 0;
  background: ${COLORS.black};
  color: ${COLORS.white};
  font: normal normal bold 14px/18px;
  padding: 14px 30px;
  border-radius: 4px;
  cursor: pointer;
  outline: none;

  :hover {
    background: ${COLORS.gray};
    color: ${COLORS.white};
  }
`;
