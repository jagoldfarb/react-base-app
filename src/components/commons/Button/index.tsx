import React, { FunctionComponent } from 'react';
import { Link } from 'react-router-dom';

import { IButtonProps } from './types';
import { CustomButton } from './styles';

const Button: FunctionComponent<IButtonProps> = (props: IButtonProps) => {
  const {
    children,
    className = '',
    disabled = false,
    href = null,
    onClick,
    type = 'button',
  } = props;

  const renderButton = () => (
    <CustomButton
      className={className}
      data-testid="button"
      disabled={disabled}
      onClick={onClick}
      type={type}
    >
      {children}
    </CustomButton>
  );

  return href ? <Link to={href}>{renderButton()}</Link> : renderButton();
};

export default Button;
