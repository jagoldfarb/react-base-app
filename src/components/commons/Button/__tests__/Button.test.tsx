import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import Button from '..';

const defaultProps = {
  children: <span>I am a button</span>,
};

describe('Button Component', () => {
  it('Should render the button without functionality', () => {
    const { asFragment, getByTestId, getByText } = render(
      <MemoryRouter>
        <Button {...defaultProps} />
      </MemoryRouter>,
    );
    expect(getByTestId('button')).toBeDefined();
    expect(getByText('I am a button')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the button with onClick function', () => {
    const props = { ...defaultProps, onClick: jest.fn() };
    const { asFragment, getByTestId, getByText } = render(
      <MemoryRouter>
        <Button {...props} />
      </MemoryRouter>,
    );
    const button = getByTestId('button');
    expect(button).toBeDefined();
    fireEvent.click(button);
    expect(getByText('I am a button')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the button as a link', () => {
    const props = { ...defaultProps, href: '/home' };
    const { asFragment, getByTestId, getByText } = render(
      <MemoryRouter>
        <Button {...props} />
      </MemoryRouter>,
    );
    const button = getByTestId('button');
    expect(button).toBeDefined();
    fireEvent.click(button);
    expect(getByText('I am a button')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });
});
