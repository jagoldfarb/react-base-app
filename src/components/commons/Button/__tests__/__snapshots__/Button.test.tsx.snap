// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Should render the button as a link 1`] = `
<DocumentFragment>
  <a
    href="/home"
  >
    <button
      class="sc-bdfBwQ ejaaIL"
      data-testid="button"
      type="button"
    >
      <span>
        I am a button
      </span>
    </button>
  </a>
</DocumentFragment>
`;

exports[`Button Component Should render the button with onClick function 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ ejaaIL"
    data-testid="button"
    type="button"
  >
    <span>
      I am a button
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button Component Should render the button without functionality 1`] = `
<DocumentFragment>
  <button
    class="sc-bdfBwQ ejaaIL"
    data-testid="button"
    type="button"
  >
    <span>
      I am a button
    </span>
  </button>
</DocumentFragment>
`;
